src/mlpack/core/optimizers/saga/saga_impl.hpp

```diff
@@ -0,0 +1,185 @@
+  double lastObjective = DBL_MAX;
+
+  // Initialize the update policy.
+  if (resetPolicy)
+    updatePolicy.Initialize(iterate.n_rows, iterate.n_cols);
+
+  // Now iterate!
+  arma::mat gradient(iterate.n_rows, iterate.n_cols);
+  arma::mat gradient0(iterate.n_rows, iterate.n_cols);
+  arma::mat iterate0;
+
+  // Find the number of batches.
+  size_t numBatches = numFunctions / batchSize;
+  if (numFunctions % batchSize != 0)
+    ++numBatches; // Capture last few.
+
+  const size_t actualMaxIterations = (maxIterations == 0) ?
+               std::numeric_limits<size_t>::max() : maxIterations;
+  for (size_t i = 0; i < actualMaxIterations; ++i)
+  {
+    // Calculate the objective function.
+    overallObjective = 0;
+    for (size_t f = 0; f < numFunctions; f += batchSize)
+    {
+      const size_t effectiveBatchSize = std::min(batchSize, numFunctions - f);
+      overallObjective += function.Evaluate(iterate, f, effectiveBatchSize);
+    }
+
+    if (std::isnan(overallObjective) || std::isinf(overallObjective))
+    {
+      Log::Warn << "SAGA: converged to " << overallObjective
+                << "; terminating  with failure.  Try a smaller step size?"
+                << std::endl;
+      return overallObjective;
+    }
+
+    if (std::abs(lastObjective - overallObjective) < tolerance)
+    {
+      Log::Info << "SAGA: minimized within tolerance " << tolerance
+                << "; terminating optimization." << std::endl;
+      return overallObjective;
+    }
+
+    lastObjective = overallObjective;
+
+    // initialize the gradients.
+    size_t effectiveBatchSize = std::min(batchSize, numFunctions);
+    arma::cube tableOfGradients(iterate.n_rows, iterate.n_cols, numBatches);
+    arma::mat avgGradient(iterate.n_rows, iterate.n_cols);
+
+    for (size_t f = 0, b=0; f < numFunctions;
+      /* incrementing done manually */)
+    {
+      // Find the effective batch size (the last batch may be smaller).
+      effectiveBatchSize = std::min(batchSize, numFunctions - f);
+
+      function.Gradient(iterate, f, gradient, effectiveBatchSize);
+      tableOfGradients.slice(b) = gradient;
+
+      f += effectiveBatchSize;
+      avgGradient += gradient;
+      b++;
+    }
+    avgGradient /= (double) numBatches;
+
+    // Store current parameter for the calculation of the variance reduced
+    // gradient.
+    iterate0 = iterate;
+
+    for (size_t f = 0, b=0, currentFunction = 0; f < numFunctions;
+      /* incrementing done manually */)
+    {
+      // Is this iteration the start of a sequence?
+      if ((currentFunction % numFunctions) == 0)
+      {
+        currentFunction = 0;
+
+        // Determine order of visitation.
+        if (shuffle)
+          function.Shuffle();
+      }
+
+      b = math::RandInt(0, numBatches); // Random batch selected
+      // Find the effective batch size (the last batch may be smaller).
+      if (b == numBatches-1)
+      {
+        effectiveBatchSize = numFunctions - b*batchSize;
```

Comment:
I think we could avoid the condition here, and use something like:

https://github.com/mlpack/mlpack/blob/130918a5b3bf961eb1f88b251ad56035e9ad7e3c/src/mlpack/core/optimizers/sgd/sgd_impl.hpp#L106-L114